### DISCUSSION TITLE
Poloniex Exceptions Handler

### DISCUSSION THREAD

**Discussion by David-Pena** [s4w3d0ff/python-poloniex#210] [Q&A]
I'm trying to handle poloniex exceptions but in https://docs.poloniex.com/ can't find anything about it.

Is there another doc should I see?

**Comment by s4w3d0ff** [comment]
What exception are you trying to handle?

This wrapper handles a select few exceptions internally. All the exceptions are handled in the [_handleReturned](https://github.com/s4w3d0ff/python-poloniex/blob/41f7048dc3be21739ba596abb27f0a2007a014ad/poloniex/__init__.py#L259) method and [_retry](https://github.com/s4w3d0ff/python-poloniex/blob/41f7048dc3be21739ba596abb27f0a2007a014ad/poloniex/__init__.py#L148) decorator.

If a `RequestsException` is raised during an api call to poloniex, the `_retry` decorator will [capture it and retry](https://github.com/s4w3d0ff/python-poloniex/blob/41f7048dc3be21739ba596abb27f0a2007a014ad/poloniex/__init__.py#L159) the call.

If an error status code is returned from poloniex, `_handleReturned` [will raise](https://github.com/s4w3d0ff/python-poloniex/blob/41f7048dc3be21739ba596abb27f0a2007a014ad/poloniex/__init__.py#L265) a `RequestsException` triggering the _retry decorator and output what was returned to the logger at the 'error' level.

If poloniex returns something other than json data the `_handleReturned` method will raise a `PoloniexError` with a ['Invalid json response returned'](https://github.com/s4w3d0ff/python-poloniex/blob/41f7048dc3be21739ba596abb27f0a2007a014ad/poloniex/__init__.py#L275) message (example: #198) and [the returned output is logged](https://github.com/s4w3d0ff/python-poloniex/blob/41f7048dc3be21739ba596abb27f0a2007a014ad/poloniex/__init__.py#L274) at the 'error' level.

Sometimes Poloniex returns an 'error' field within the json data. This wrapper looks for "_Nonce must be greater_", attempts to update the nonce and raises a `RequestsException` to trigger the `_retry` decorator and try the call again. This wrapper also looks for "_please try again_" and again raises a `RequestsException` to trigger `_retry` ([see here](https://github.com/s4w3d0ff/python-poloniex/blob/41f7048dc3be21739ba596abb27f0a2007a014ad/poloniex/__init__.py#L277-L290)).

Any other message in the 'error' field is [captured and raised](https://github.com/s4w3d0ff/python-poloniex/blob/41f7048dc3be21739ba596abb27f0a2007a014ad/poloniex/__init__.py#L294) as a `PoloniexError`.

The only [other exception](https://github.com/s4w3d0ff/python-poloniex/blob/41f7048dc3be21739ba596abb27f0a2007a014ad/poloniex/__init__.py#L163) this wrapper should raise is a `RetryException` which has a list of all the `RequestsException`s captured while attempting to retry api calls.

Most errors returned will be logged by the `poloniex.logger` at the error level. So when debugging you can use the logging module to see the output:
```python
import logging

logging.basicConfig()
poloniex.logger.setLevel(logging.DEBUG)
```

Hope this helps.
